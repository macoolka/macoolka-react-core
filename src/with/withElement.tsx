import * as React from 'react';
import { ElementProps } from '../types';
import { Props } from '../css'
import { getWrapDisplayName, getWrapedDefaultProps } from '../reactHelper';
import { omit } from 'macoolka-object'
const withElement = <P extends React.HTMLAttributes<any> & Props>
    (WrapedComponent: React.ComponentType<P &{ref?:any}>) => {
    const C = React.forwardRef((props: P & ElementProps,ref:any) => {

        return (
            <WrapedComponent
                ref={ref}
                {...omit(props, ['tabIndex', 'onClick']) as any}
                tabIndex={props.disabled ? -1 : props.tabIndex}
                onClick={props.disabled ? undefined : props.onClick}
                mkTextColor={props.disabled ? 'disabled' : props.mkTextColor}
                mkHoverTextColor={props.disabled ? undefined : props.mkHoverTextColor}
                mkHoverColor={props.disabled ? undefined : props.mkHoverColor}
                mkColor={props.disabled ? undefined : props.mkColor}
            />
        );

    })
    C.displayName = getWrapDisplayName('withElement')(WrapedComponent)
    // const originDefaultProps:{}= WrapedComponent.defaultProps? WrapedComponent.defaultProps:{};
    C.defaultProps = getWrapedDefaultProps(WrapedComponent)({ tabIndex: 0 } as any)
    /*   
        C.defaultProps = {
            tabIndex: 0,
            ...originDefaultProps
        } as ElementProps & P */
    return C;
}
export default withElement;
