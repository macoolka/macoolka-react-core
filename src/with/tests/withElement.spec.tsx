import * as React from 'react';
/* import ReactDOM from 'react-dom';
import keycode from 'keycode'; */
import { shallow } from 'enzyme';
import withElement from '../withElement';
import renderer from 'react-test-renderer'
import 'jest-styled-components'
describe('<withElement />', () => {

    const _A: React.SFC<React.ButtonHTMLAttributes<{}>> = ({ children, ...others }) => (<button {...others}>{children}</button>)
    const A = withElement(_A);

    describe('root node', () => {
        it('should render a element with type="div" by default', () => {
            const wrapper = shallow(<A>Hello</A>);
            expect(wrapper.name()).toEqual('_A');
            expect((wrapper.dive() as any).props().children).toEqual('Hello');
            expect(wrapper.props().type).toBeUndefined();
            const tree = renderer.create(<A>Hello</A>).toJSON()
            expect(tree).toMatchSnapshot()
        });
        it('should change the  type', () => {
            const R = <A type="submit">Hello</A>
            const wrapper = shallow(R);
            expect(wrapper.props().type).toEqual('submit');
            const tree = renderer.create(R).toJSON()
            expect(tree).toMatchSnapshot()
        });

        it('should change the element component and add accessibility requirements', () => {
            const R = <A role="checkbox" aria-checked={false}>Hello</A>
            const wrapper = shallow(R);
            expect(wrapper.props().role).toEqual('checkbox');
            expect(wrapper.props().tabIndex).toEqual(0);
            const tree = renderer.create(R).toJSON()
            expect(tree).toMatchSnapshot()

        });
        it('should spread props on button', () => {
            const R = <A data-test="hello">Hello</A>
            const wrapper = shallow(R);
            expect(wrapper.prop('data-test')).toEqual('hello');
            const tree = renderer.create(R).toJSON()
            expect(tree).toMatchSnapshot()
        });
        it('should spread props on button', () => {
            const R = <A disabled tabIndex={100}>Hello</A>
            const wrapper = shallow(R);
            expect(wrapper.prop('tabIndex')).toEqual(-1);
            const tree = renderer.create(R).toJSON()
            expect(tree).toMatchSnapshot()
        });
    });


});
