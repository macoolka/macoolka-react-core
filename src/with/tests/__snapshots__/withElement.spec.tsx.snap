// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withElement /> root node should change the  type 1`] = `
<button
  tabIndex={0}
  type="submit"
>
  Hello
</button>
`;

exports[`<withElement /> root node should change the element component and add accessibility requirements 1`] = `
<button
  aria-checked={false}
  role="checkbox"
  tabIndex={0}
>
  Hello
</button>
`;

exports[`<withElement /> root node should render a element with type="div" by default 1`] = `
<button
  tabIndex={0}
>
  Hello
</button>
`;

exports[`<withElement /> root node should spread props on button 1`] = `
<button
  data-test="hello"
  tabIndex={0}
>
  Hello
</button>
`;

exports[`<withElement /> root node should spread props on button 2`] = `
<button
  disabled={true}
  mkTextColor="disabled"
  tabIndex={-1}
>
  Hello
</button>
`;
